eth/api_tracer_oe.go

```diff
@@ -0,0 +1,181 @@
+	Result              interface{}       `json:"result"`
+	Subtraces           int               `json:"subtraces"`
+	TraceAddress        []int             `json:"traceAddress"`
+	TransactionHash     *common.Hash      `json:"transactionHash"`
+	TransactionPosition *uint64           `json:"transactionPosition"`
+	Type                string            `json:"type"`
+}
+
+// TraceRewardAction An OpenEthereum formatted trace reward action
+type TraceRewardAction struct {
+	Value      *hexutil.Big    `json:"value,omitempty"`
+	Author     *common.Address `json:"author,omitempty"`
+	RewardType string          `json:"rewardType,omitempty"`
+}
+
+// setConfigTracerToOpenEthereum forces the Tracer to the OpenEthereum one, mutating the argument value.
+func setConfigTracerToOpenEthereum(config *TraceConfig) *TraceConfig {
+	if config == nil {
+		config = &TraceConfig{}
+	}
+
+	tracer := "callTracerOpenEthereum"
+	config.Tracer = &tracer
+	return config
+}
+
+func traceBlockReward(ctx context.Context, eth *Ethereum, block *types.Block, config *TraceConfig) (*OpenEthereumTrace, error) {
+	chainConfig := eth.blockchain.Config()
+	minerReward, _ := ethash.AccumulateRewards(chainConfig, block.Header(), block.Uncles())
+
+	coinbase := block.Coinbase()
+	blockHash := block.Hash()
+
+	tr := &OpenEthereumTrace{
+		Type: "reward",
+		Action: TraceRewardAction{
+			Value:      (*hexutil.Big)(minerReward),
+			Author:     &coinbase,
+			RewardType: "block",
+		},
+		TraceAddress: []int{},
+		BlockHash:    &blockHash,
+		BlockNumber:  block.NumberU64(),
+	}
+
+	return tr, nil
+}
+
+func traceBlockUncleRewards(ctx context.Context, eth *Ethereum, block *types.Block, config *TraceConfig) ([]*OpenEthereumTrace, error) {
+	chainConfig := eth.blockchain.Config()
+	_, uncleRewards := ethash.AccumulateRewards(chainConfig, block.Header(), block.Uncles())
+
+	blockHash := block.Hash()
+
+	results := make([]*OpenEthereumTrace, len(uncleRewards))
+	for i, uncle := range block.Uncles() {
+		if i < len(uncleRewards) {
+			coinbase := uncle.Coinbase
+
+			results[i] = &OpenEthereumTrace{
+				Type: "reward",
+				Action: TraceRewardAction{
+					Value:      (*hexutil.Big)(uncleRewards[i]),
+					Author:     &coinbase,
+					RewardType: "uncle",
+				},
+				TraceAddress: []int{},
+				BlockNumber:  block.NumberU64(),
+				BlockHash:    &blockHash,
+			}
+		}
+	}
+
+	return results, nil
+}
+
+// Block returns the structured logs created during the execution of
+// EVM and returns them as a JSON object.
+// The correct name will be TraceBlockByNumber, though we want to be compatible with OpenEthereum trace module.
+func (api *PrivateTraceAPI) Block(ctx context.Context, number rpc.BlockNumber, config *TraceConfig) ([]interface{}, error) {
+	// Fetch the block that we want to trace
+	var block *types.Block
+
+	switch number {
+	case rpc.PendingBlockNumber:
+		block = api.eth.miner.PendingBlock()
+	case rpc.LatestBlockNumber:
+		block = api.eth.blockchain.CurrentBlock()
```

Comment:
Should we handle `case rpc.EarliestBlockNumber` here too?

Reply:
Probably not, I copied this from the `debug_traceBlock` where it's neither being handled. The reason it's not being handled is, that we can't retrieve block trace for the genesis block. We can add it if a reason exists, though the current logic fails when it tries to read the state of the parent block, as genesis has no parent block.

p.s.: I have extracted the logic of getting a block from a block number in a stash I have for `trace_filter` support.